src/analysis/complex/upper_half_plane.lean

```diff
@@ -84,79 +88,149 @@ begin
   ring
 end
 
-lemma mul_smul' (x y : SL(2, ℝ)) (z : ℍ) :
+lemma mul_smul' (x y :  GL(2, ℝ)⁺) (z : ℍ) :
   smul_aux (x * y) z = smul_aux x (smul_aux y z) :=
 begin
   ext1,
   change _ / _ = (_ * (_ / _) + _)  * _,
   rw denom_cocycle,
   field_simp [denom_ne_zero, -denom, -num],
-  simp [matrix.mul, dot_product, fin.sum_univ_succ],
+  simp only [matrix.mul, dot_product, fin.sum_univ_succ, num, denom, coe_coe, subgroup.coe_mul,
+  general_linear_group.coe_mul, fintype.univ_of_subsingleton, fin.mk_eq_subtype_mk, fin.mk_zero,
+  finset.sum_singleton, fin.succ_zero_eq_one,complex.of_real_add, complex.of_real_mul],
   ring
 end
 
-/-- The action of `SL(2, ℝ)` on the upper half-plane by fractional linear transformations. -/
-instance : mul_action SL(2, ℝ) ℍ :=
+/-- The action of ` GL_pos 2 ℝ` on the upper half-plane by fractional linear transformations. -/
+instance : mul_action  (GL(2, ℝ)⁺) ℍ :=
 { smul := smul_aux,
-  one_smul := λ z, by { ext1, change _ / _ = _, simp },
+  one_smul := λ z, by { ext1, change _ / _ = _,
+   simp [coe_fn_coe_base']  },
   mul_smul := mul_smul' }
 
-@[simp] lemma coe_smul (g : SL(2, ℝ)) (z : ℍ) : ↑(g • z) = num g z / denom g z := rfl
-@[simp] lemma re_smul (g : SL(2, ℝ)) (z : ℍ) : (g • z).re = (num g z / denom g z).re := rfl
+section modular_scalar_towers
 
-lemma im_smul (g : SL(2, ℝ)) (z : ℍ) : (g • z).im = (num g z / denom g z).im := rfl
+variable (Γ : subgroup (special_linear_group (fin 2) ℤ))
 
-lemma im_smul_eq_div_norm_sq (g : SL(2, ℝ)) (z : ℍ) :
-  (g • z).im = z.im / (complex.norm_sq (denom g z)) :=
-smul_aux'_im g z
+instance SL_action {R : Type*} [comm_ring R] [algebra R ℝ] : mul_action SL(2, R) ℍ :=
+ mul_action.comp_hom ℍ  (monoid_hom.comp (special_linear_group.to_GL_pos)
+ (map (algebra_map R ℝ)) )
+
+instance : has_coe SL(2,ℤ) (GL(2, ℝ)⁺) := ⟨λ g , ((g : SL(2, ℝ)) : (GL(2, ℝ)⁺))⟩
+
+instance SL_on_GL_pos : has_scalar SL(2,ℤ) (GL(2, ℝ)⁺) := ⟨λ s g, s * g⟩
+
+lemma SL_on_GL_pos_smul_apply (s : SL(2,ℤ)) (g : (GL(2, ℝ)⁺) ) (z : ℍ) :
+  (s • g) • z = ( (s : GL(2, ℝ)⁺) * g) • z := rfl
+
+instance SL_to_GL_tower : is_scalar_tower SL(2,ℤ) (GL(2, ℝ)⁺) ℍ :=
+ {smul_assoc := by {intros s g z, simp only [SL_on_GL_pos_smul_apply, coe_coe], apply mul_smul',},}
+
+instance subgroup_GL_pos : has_scalar Γ (GL(2, ℝ)⁺) := ⟨λ s g, s * g⟩
+
+lemma subgroup_on_GL_pos_smul_apply (s : Γ) (g : (GL(2, ℝ)⁺) ) (z : ℍ) :
+  (s • g) • z = ( (s : GL(2, ℝ)⁺) * g) • z := rfl
+
+instance subgroup_on_GL_pos : is_scalar_tower Γ (GL(2, ℝ)⁺) ℍ :=
+ {smul_assoc :=
+  by {intros s g z, simp only [subgroup_on_GL_pos_smul_apply, coe_coe], apply mul_smul',},}
+
+instance subgroup_SL : has_scalar Γ SL(2,ℤ) := ⟨λ s g, s * g⟩
+
+lemma subgroup_on_SL_apply (s : Γ) (g : SL(2,ℤ) ) (z : ℍ) :
+  (s • g) • z = ( (s : SL(2, ℤ)) * g) • z := rfl
+
+instance subgroup_to_SL_tower : is_scalar_tower Γ SL(2,ℤ) ℍ :=
+ {smul_assoc := by {intros s g z, rw subgroup_on_SL_apply, apply upper_half_plane.SL_action.3,},}
```

Comment:
Indents again
```suggestion
instance subgroup_to_SL_tower : is_scalar_tower Γ SL(2,ℤ) ℍ :=
{ smul_assoc := λ s g z, by { rw subgroup_on_SL_apply, apply upper_half_plane.SL_action.3 } }
```
What is `upper_half_plane.SL_action.3`?

Reply:
Its the `mul_smul` given by the `SL_action`. I could use  `upper_half_plane.SL_action.mul_smul` if preferable?

Reply:
Its the `mul_smul` given by the `SL_action`. I could use  `upper_half_plane.SL_action.mul_smul` if preferable?

Reply:
Does `mul_smul` by itself not work?

Reply:
oh yes! I've changed to that now